### PR TITLE
introduce Eth2Hash, Eth2Digest and friends

### PR DESCRIPTION
**Pull request by arnetheduck**

### CODE REVIEW

**Comment by tersec** [comment]
This is necessary type abstraction.

File: beacon_chain/datatypes.nim [hunk 1]
```diff
@@ -133,26 +132,26 @@ type
     ## worth of assignments
     persistent_committees*: seq[seq[ValidatorRecord]]      # Persistent shard committees
     persistent_committee_reassignments*: seq[ShardReassignmentRecord]
-    next_shuffling_seed*: Blake2_256_Digest                # Randao seed used for next shuffling
+    next_shuffling_seed*: Eth2Digest                # Randao seed used for next shuffling
     deposits_penalized_in_period*: uint32                  # Total deposits penalized in the given withdrawal period
```
Comment (tersec) [inline beacon_chain/datatypes.nim]: Here, as elsewhere, the presumably mechanical search/replace process didn't account for `Eth2Digest` having 7 fewer characters than `Blake2_256_Digest` and thus disrupting the comment alignment.

Comment (arnetheduck) [reply beacon_chain/datatypes.nim]: bah. `<insert rant about the hopelessness of manually aligned comments here>`

File: beacon_chain/ssz.nim [hunk 2]
```diff
@@ -11,8 +11,8 @@
 
 import
   endians, typetraits, options, algorithm,
-  eth_common, nimcrypto,
-  ./datatypes
+  eth_common, nimcrypto/blake2,
```
Comment (tersec) [inline beacon_chain/ssz.nim]: I see why it's necessary to import this given how the rest of `ssz.nim` is set up, but it seem to be a too-leaky abstraction. Better than before, but still not ideal.

In general, this PR is mostly mechanical, making almost no semantic changes, which is useful -- small architectural issues such as this are relatively fine insofar as the improvements far outweigh them.

Comment (arnetheduck) [reply beacon_chain/ssz.nim]: same reason as with init - there's some lookup issue with generics and mixins and whatnot that @zah and @mratsim are bringing up with @Araq, that can be worked around with these extra includes.

**Pull request merged**